File: server/apps/cybersecurity/schemas/shared/feedback_schemas.py
"""
Feedback Schemas — in-app bug reports, feature requests, questions.

Field names match the Feedback model (models/feedback_models.py) exactly. `created_by`
(submitter) is enriched on responses with the user's email/name (JOIN on accounts.User).
"""

from enum import Enum
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


# ==========================================
# Enums
# ==========================================

class FeedbackCategory(str, Enum):
    BUG = "bug"
    IMPROVEMENT = "improvement"
    FEATURE = "feature"
    QUESTION = "question"
    OTHER = "other"


class FeedbackStatus(str, Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    CLOSED = "closed"


# ==========================================
# Request Schemas
# ==========================================

class FeedbackCreate(BaseModel):
    """Schema for submitting feedback (any authenticated user)."""
    category: FeedbackCategory = Field(description="bug | improvement | feature | question | other")
    title: str = Field(min_length=3, max_length=200)
    description: str = Field(min_length=10)
    page_url: str | None = Field(default=None, description="Page the feedback was submitted from")


class FeedbackContentUpdate(BaseModel):
    """Schema for the submitter (or admin) editing the feedback content (not triage)."""
    category: FeedbackCategory | None = None
    title: str | None = Field(default=None, min_length=3, max_length=200)
    description: str | None = Field(default=None, min_length=10)


class FeedbackUpdate(BaseModel):
    """Schema for admin triage — status + internal notes only."""
    status: FeedbackStatus | None = None
    admin_notes: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class FeedbackDetail(BaseModel):
    """Schema for feedback details."""
    id: int
    category: str
    status: str
    title: str
    description: str
    page_url: str | None = None
    admin_notes: str | None = None
    organization_id: int | None = None
    created_by: int | None = None
    # Resolved from accounts.User on created_by so the admin sees who submitted
    # without a second request — enriched-schema pattern.
    user_email: str | None = None
    user_name: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class FeedbackListResponse(BaseModel):
    """Response schema for listing feedback."""
    success: bool
    data: list[FeedbackDetail] | None = None
    count: int = 0
    error: str | None = None


class FeedbackResponse(BaseModel):
    """Response schema for single feedback operations."""
    success: bool
    data: FeedbackDetail | None = None
    error: str | None = None


class FeedbackMessageResponse(BaseModel):
    """Simple message response."""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/cybersecurity/schemas/shared/feedback.schemas.ts
import { z } from 'zod';

// ==========================================
// Enums (mirror backend feedback_schemas.py)
// ==========================================

// Why strict enums: prevent casing/typo bugs and keep parity with the Pydantic enums.
export const FeedbackCategorySchema = z.enum(['bug', 'improvement', 'feature', 'question', 'other']);
export const FeedbackStatusSchema = z.enum(['open', 'in_progress', 'resolved', 'closed']);

// ==========================================
// Full Schema (GET responses)
// ==========================================

// Matches backend FeedbackDetail. user_email/user_name are enriched server-side from
// accounts.User on created_by (so the admin sees who submitted without a second request).
export const FeedbackSchema = z.object({
  id: z.number(),
  category: FeedbackCategorySchema,
  status: FeedbackStatusSchema,
  title: z.string(),
  description: z.string(),
  page_url: z.string().nullable(),
  admin_notes: z.string().nullable(),
  organization_id: z.number().nullable(),
  created_by: z.number().nullable(),
  user_email: z.string().nullable(),
  user_name: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

// Submit feedback (any authenticated user). page_url is auto-captured by the page.
export const CreateFeedbackSchema = z.object({
  category: FeedbackCategorySchema,
  title: z.string().min(3).max(200),
  description: z.string().min(10),
  page_url: z.string().nullable().optional(),
});

// Owner/admin content edit — category/title/description (mirrors backend FeedbackContentUpdate).
export const UpdateFeedbackContentSchema = z.object({
  category: FeedbackCategorySchema.optional(),
  title: z.string().min(3).max(200).optional(),
  description: z.string().min(10).optional(),
});

// Admin triage only — status + internal notes (mirrors backend FeedbackUpdate).
export const UpdateFeedbackSchema = z.object({
  status: FeedbackStatusSchema.optional(),
  admin_notes: z.string().nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type FeedbackCategory = z.infer<typeof FeedbackCategorySchema>;
export type FeedbackStatus = z.infer<typeof FeedbackStatusSchema>;
export type Feedback = z.infer<typeof FeedbackSchema>;
export type CreateFeedback = z.infer<typeof CreateFeedbackSchema>;
export type UpdateFeedbackContent = z.infer<typeof UpdateFeedbackContentSchema>;
export type UpdateFeedback = z.infer<typeof UpdateFeedbackSchema>;

// ==========================================
// Labels (UI display)
// ==========================================

export const FEEDBACK_CATEGORY_LABELS: Record<FeedbackCategory, string> = {
  bug: 'Bug',
  improvement: 'Improvement',
  feature: 'Feature Request',
  question: 'Question',
  other: 'Other',
};

export const FEEDBACK_STATUS_LABELS: Record<FeedbackStatus, string> = {
  open: 'Open',
  in_progress: 'In Progress',
  resolved: 'Resolved',
  closed: 'Closed',
};

export const getFeedbackCategoryLabel = (category: string): string =>
  FEEDBACK_CATEGORY_LABELS[category as FeedbackCategory] || category;

export const getFeedbackStatusLabel = (status: string): string =>
  FEEDBACK_STATUS_LABELS[status as FeedbackStatus] || status;

// ==========================================
// Response Types (FastAPI wrapper { success, data, count, error })
// ==========================================

export type FeedbackResponse = {
  success: boolean;
  data?: Feedback;
  error?: string;
};

export type FeedbacksResponse = {
  success: boolean;
  data?: Feedback[];
  count?: number;
  error?: string;
};

export type FeedbackMessageResponse = {
  success: boolean;
  message?: string;
  error?: string;
};
